### PR TITLE
🏗 Quickly stop experiment builds before checking them out if they are not configured on the tested branch

### PR DESCRIPTION
**Pull request by danielrozenberg**
See examples for debugging commit 0914d35:
* Experiment A was set to be [expired](https://app.circleci.com/pipelines/github/ampproject/amphtml/6631/workflows/4296ede8-140e-4d09-b5a6-c54eb67b3764/jobs/95258)
* Experiment B is valid and [gets built](https://app.circleci.com/pipelines/github/ampproject/amphtml/6631/workflows/4296ede8-140e-4d09-b5a6-c54eb67b3764/jobs/95259)
* Experiment C is [not configured](https://app.circleci.com/pipelines/github/ampproject/amphtml/6631/workflows/4296ede8-140e-4d09-b5a6-c54eb67b3764/jobs/95253)

### CODE REVIEW

File: .circleci/maybe_gracefully_halt.sh [hunk 1]
```diff
@@ -21,8 +21,36 @@
...
+  exit 0
 fi
+
+if [[ $CIRCLE_JOB == Experiment* ]]; then
+  EXP=$(echo $CIRCLE_JOB | awk '{print $2}')
+  if [[ -f /tmp/restored-workspace/.CIRCLECI_MERGE_COMMIT ]]; then
+    COMMIT_SHA="$(cat /tmp/restored-workspace/.CIRCLECI_MERGE_COMMIT)"
+  else
+    COMMIT_SHA="${CIRCLE_SHA1}"
+  fi
```
Comment (rsimha) [inline .circleci/maybe_gracefully_halt.sh:38]:
```suggestion
  # Extract the experiment name from the job name in `config.yml`.
  EXP=$(echo $CIRCLE_JOB | awk '{print $2}')
  
  # Extract the commit SHA. For PR jobs, this is written to .CIRCLECI_MERGE_COMMIT.
  if [[ -f /tmp/restored-workspace/.CIRCLECI_MERGE_COMMIT ]]; then
    COMMIT_SHA="$(cat /tmp/restored-workspace/.CIRCLECI_MERGE_COMMIT)"
  else
    COMMIT_SHA="${CIRCLE_SHA1}"
  fi
```

File: .circleci/maybe_gracefully_halt.sh [hunk 2]
```diff
@@ -21,8 +21,36 @@
 set -e
 
 GREEN() { echo -e "\n\033[0;32m$1\033[0m"; }
+YELLOW() { echo -e "\n\033[0;33m$1\033[0m"; }
 
 if ls /tmp/restored-workspace/.CI_GRACEFULLY_HALT_* 1>/dev/null 2>&1; then
   echo $(GREEN "Gracefully halting this job.")
   circleci-agent step halt
+  exit 0
```
Comment (rsimha) [inline .circleci/maybe_gracefully_halt.sh:29]:
```suggestion
# Halt the job if the well-known sentinel file was created by the previous CI stage.
if ls /tmp/restored-workspace/.CI_GRACEFULLY_HALT_* 1>/dev/null 2>&1; then
  echo $(GREEN "Gracefully halting this job.")
  circleci-agent step halt
  exit 0
```

File: .circleci/maybe_gracefully_halt.sh [hunk 3]
```diff
@@ -21,8 +21,36 @@
...
+    COMMIT_SHA="${CIRCLE_SHA1}"
+  fi
+
+  EXPERIMENT_JSON=$(curl -sS "https://raw.githubusercontent.com/ampproject/amphtml/${COMMIT_SHA}/build-system/global-configs/experiments-config.json" | jq ".experiment${EXP}")
+  if ! echo "${EXPERIMENT_JSON}" | jq -e '.name,.define_experiment_constant,.expiration_date_utc'; then
+    echo $(YELLOW "Experiment ${EXP} is misconfigured, or does not exist.")
+    echo $(GREEN "Gracefully halting this job")
+    circleci-agent step halt
+    exit 0
+  fi
```
Comment (rsimha) [inline .circleci/maybe_gracefully_halt.sh:46]:
```suggestion
  # Do not proceed if the experiment config is missing a valid name, constant, or date.
  EXPERIMENT_JSON=$(curl -sS "https://raw.githubusercontent.com/ampproject/amphtml/${COMMIT_SHA}/build-system/global-configs/experiments-config.json" | jq ".experiment${EXP}")
  if ! echo "${EXPERIMENT_JSON}" | jq -e '.name,.define_experiment_constant,.expiration_date_utc'; then
    echo $(YELLOW "Experiment ${EXP} is misconfigured, or does not exist.")
    echo $(GREEN "Gracefully halting this job")
    circleci-agent step halt
    exit 0
  fi
```

File: .circleci/maybe_gracefully_halt.sh [hunk 4]
```diff
@@ -21,8 +21,36 @@
...
+
+  CURRENT_TIMESTAMP=$(date --utc +'%s')
+  EXPERIMENT_EXPIRATION_TIMESTAMP=$(date --utc --date $(echo "${EXPERIMENT_JSON}" | jq -er '.expiration_date_utc') +'%s')
+  if [[ $CURRENT_TIMESTAMP -gt $EXPERIMENT_EXPIRATION_TIMESTAMP ]]; then
+    echo $(YELLOW "Experiment ${EXP} is expired.")
+    echo $(GREEN "Gracefully halting this job")
+    circleci-agent step halt
+    exit 0
+  fi
```
Comment (rsimha) [inline .circleci/maybe_gracefully_halt.sh:55]:
```suggestion
  # Do not proceed if the experiment is expired (config date is in the past).
  CURRENT_TIMESTAMP=$(date --utc +'%s')
  EXPERIMENT_EXPIRATION_TIMESTAMP=$(date --utc --date $(echo "${EXPERIMENT_JSON}" | jq -er '.expiration_date_utc') +'%s')
  if [[ $CURRENT_TIMESTAMP -gt $EXPERIMENT_EXPIRATION_TIMESTAMP ]]; then
    echo $(YELLOW "Experiment ${EXP} is expired.")
    echo $(GREEN "Gracefully halting this job")
    circleci-agent step halt
    exit 0
  fi
```

**Review by rsimha — APPROVED** [review]
LGTM with some comments / suggestions below.

**Some back-of-the-envelope math:** We run ~1500 CI builds a month, so ~18,000 a year. If we assume that each experiment is enabled only half the time, and if the build job for an invalid experiment was running ~3 minutes of unnecessary work earlier, this PR will save ~80,000 XLarge CPU minutes a year.

**Two high-level comments:**
- Given that these shell scripts do CI-critical stuff, I think it helps to document each paragraph with a summary comment. I've added suggestions inline. Feel free to directly commit, or reword as you'd like.
- There is now a bunch of useless code in `experiment-build.js` and `experiment-tests.js` that should probably be cleaned up. (Okay to do this in a follow up PR.)
https://github.com/ampproject/amphtml/blob/69c41e1fe1f4eafffb53c0970c3a1b0b71e3add3/build-system/pr-check/experiment-build.js#L34-L46
https://github.com/ampproject/amphtml/blob/69c41e1fe1f4eafffb53c0970c3a1b0b71e3add3/build-system/pr-check/experiment-tests.js#L61-L71

**Comment by danielrozenberg** [comment]
> this PR will save ~80,000 XLarge CPU minutes a year.

Yay!

> * Given that these shell scripts do CI-critical stuff, I think it helps to document each paragraph with a summary comment.

Merged your suggestions

> * There is now a bunch of useless code in `experiment-build.js` and `experiment-tests.js` that should probably be cleaned up.

Is there harm in keeping that code around? I can add a comment that it's duplicated in the shell files

**Comment by rsimha** [comment]
> Is there harm in keeping that code around? I can add a comment that it's duplicated in the shell files

My concern is that the code could diverge (and worse, break) if one of the two sources of truth is modified by someone else in future. Here are the actual changes I was suggesting:

**`experiment-build.js`**:

```diff
diff --git a/build-system/pr-check/experiment-build.js b/build-system/pr-check/experiment-build.js
index 171e842e9..ca7380f5e 100644
--- a/build-system/pr-check/experiment-build.js
+++ b/build-system/pr-check/experiment-build.js
@@ -33,16 +33,9 @@ const jobName = `${experiment}-build.js`;
 
 function pushBuildWorkflow() {
   const config = getExperimentConfig(experiment);
-  if (config) {
-    const defineFlag = `--define_experiment_constant ${config.define_experiment_constant}`;
-    timedExecOrDie(`amp dist --fortesting ${defineFlag}`);
-    uploadExperimentOutput(experiment);
-  } else {
-    skipDependentJobs(
-      jobName,
-      `${experiment} is expired, misconfigured, or does not exist`
-    );
-  }
+  const defineFlag = `--define_experiment_constant ${config.define_experiment_constant}`;
+  timedExecOrDie(`amp dist --fortesting ${defineFlag}`);
+  uploadExperimentOutput(experiment);
 }
 
 function prBuildWorkflow() {
```

**`experiment-tests.js`**:

```diff
diff --git a/build-system/pr-check/experiment-tests.js b/build-system/pr-check/experiment-tests.js
index 4481afadc..364cc4a55 100644
--- a/build-system/pr-check/experiment-tests.js
+++ b/build-system/pr-check/experiment-tests.js
@@ -60,14 +60,7 @@ function runExperimentTests(config) {
 
 function pushBuildWorkflow() {
   const config = getExperimentConfig(experiment);
-  if (config) {
-    runExperimentTests(config);
-  } else {
-    skipDependentJobs(
-      jobName,
-      `${experiment} is expired, misconfigured, or does not exist`
-    );
-  }
+  runExperimentTests(config);
 }
 
 function prBuildWorkflow() {
```

File: .circleci/maybe_gracefully_halt.sh (hunk 4 above)
Comment (rsimha) [reply .circleci/maybe_gracefully_halt.sh:59]: Looks like this suggestion didn't get applied. (Maybe I did it wrong?)

Comment (danielrozenberg) [reply .circleci/maybe_gracefully_halt.sh:60]: Weird. Done.

**Comment by danielrozenberg** [comment]
> My concern is that the code could diverge (and worse, break) if one of the two sources of truth is modified by someone else in future. Here are the actual changes I was suggesting:

Applied your suggestion with a comment explaining that the config is assumed to be valid.

**Comment by danielrozenberg** [comment]
Note that this PR doesn't pass CircleCI tests now because it uses the shell script from `main`, and therefore doesn't skip Experiment C, which is not configured. You'll have to force-merge it

**Comment by rsimha** [comment]
> script from `main`, and therefore doesn't skip Experiment C, which is not configured. You'll have to force-merge it

Yep, I was about to say the same thing. Would it be possible to create a test build with a commit that `curl`s the file from your PR branch (to make sure everything works as desired)? After that, I can force merge this version that points to `ampproject/amphtml/main`.

**Review by rsimha — APPROVED** [review]

**Comment by danielrozenberg** [comment]
Done, let's let it run

**Comment by rsimha** [comment]
> Done, let's let it run

Yay! ([link](https://app.circleci.com/pipelines/github/ampproject/amphtml/6800/workflows/f16044e4-58b6-4600-8adf-0c319f69c326))

image

**Comment by danielrozenberg** [comment]
Let's let run to completion and then I'll revert the last commit

**Comment by danielrozenberg** [comment]
Ready!

**Pull request merged**